#!/usr/bin/bash

##########################################################
# Genome Analysis Script for Breast Cancer Research
# Author: Vikas Kaushik
# Date: 2023
# Description: This script performs various analyses including FastQC, HISAT2 alignment, variant calling, 
# normalization, and variant comparison between test and control samples (Breast Cancer).
##########################################################

# Bash Script for FastQC which generates an HTML report that can be easily interpreted and visually inspected.

# Define the list of sample names
SAMPLES=("SRR19641513" "SRR19641514" "SRR19641515" "SRR19641516" "SRR19641517" "SRR19641518")

# Run FastQC for each sample
for SAMPLE in "${SAMPLES[@]}"; do
	        fastqc "$INPUT_DIR/${SAMPLE}_F.gz" "$INPUT_DIR/${SAMPLE}_R.gz" -o "$OUTPUT_DIR"
done

# Bash script for HISAT2 alignment
for SAMPLE in "${SAMPLES[@]}"; do
    hisat2 -p 8 --dta -x "indexes/Chromosome17_index" -1 "samples/${SAMPLE}_F.gz" -2 "samples/${SAMPLE}_R.gz" -S "map/${SAMPLE}.sam" --rg-id "${SAMPLE}" --rg "ID:${SAMPLE}" --rg "CN:BCCAGSC" --rg "LB:NA" --rg "PL:illumina" --rg "PU:NA" --rg "SM:${SAMPLE}"
    
    samtools view -bS "map/${SAMPLE}.sam" > "map/${SAMPLE}.bam"
    samtools sort "map/${SAMPLE}.bam" -o "map/${SAMPLE}_sorted.bam"
    samtools index "map/${SAMPLE}_sorted.bam"
done

 # For removal of duplicates
    for SAMPLE in "${SAMPLES[@]}"; do
    samtools rmdup "map/${SAMPLE}_sorted.bam" "map/${SAMPLE}_deduplicated.bam"
done

# Bash script for variant calling and filtering

# Loop over each sample
for SAMPLE in "${SAMPLES[@]}"; do
    # Variant calling with bcftools mpileup and bcftools call
    bcftools mpileup -Ou -f "reference/Chromosome17.fasta" "map/${SAMPLE}_deduplicated.bam" | bcftools call -mv -O v -o "variants/${SAMPLE}.vcf"
done

# # Normalization for sall samples
for SAMPLE in "${SAMPLES[@]}"; do
    bcftools norm -f "reference/Chromosome17.fasta" -O v -o "normalized/${SAMPLE}_normalized.vcf" "variants/${SAMPLE}.vcf"
    done

# Joint genotyping for test samples
bcftools mpileup -Ou -f reference/Chromosome17.fasta map/SRR19641513_sorted.bam map/SRR19641514_sorted.bam map/SRR19641515_sorted.bam | bcftools call -mv -Ov -o variants/test_joint_variants.vcf
# Normalization for test samples
bcftools norm -f "reference/Chromosome17.fasta" -O v -o "normalized/test_joint_variants_normalized.vcf" "variants/test_joint_variants.vcf"

# Joint genotyping for control samples
bcftools mpileup -Ou -f reference/Chromosome17.fasta map/SRR19641516_sorted.bam map/SRR19641517_sorted.bam map/SRR19641518_sorted.bam | bcftools call -mv -Ov -o variants/control_joint_variants.vcf
# Normalization for control samples
bcftools norm -f "reference/Chromosome17.fasta" -O v -o "normalized/control_joint_variants_normalized.vcf" "variants/control_joint_variants.vcf"

# Compress VCF files using bgzip and then index them with tabix
bgzip normalized/test_joint_variants_normalized.vcf
bgzip normalized/control_joint_variants_normalized.vcf
tabix -p vcf normalized/test_joint_variants_normalized.vcf.gz
tabix -p vcf normalized/control_joint_variants_normalized.vcf.gz

# Compare variants between test and control
bcftools isec -p variants -O z normalized/test_joint_variants_normalized.vcf.gz normalized/control_joint_variants_normalized.vcf.gz

# The files named 0000.vcf.gz, 0001.vcf.gz, 0002.vcf.gz, and 0003.vcf.gz are output files generated by the bcftools isec command. These files typically represent subsets of variants based on their intersection or differences between test and control  VCF files.
# 0000.vcf.gz: Variants that are common (intersection) between the test and control samples.
# 0001.vcf.gz: Variants unique to the test sample.
# 0002.vcf.gz: Variants unique to the control sample.
# 0003.vcf.gz: Variants that are different between the test and control samples.
